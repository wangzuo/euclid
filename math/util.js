const { tan, acos, sqrt } = Math;

// min (inclusive) and max (exclusive)
export const random = (min, max) => Math.random() * (max - min) + min;

export const dist = ({ x: x1, y: y1 }, { x: x2, y: y2 }) => {
  return Math.sqrt((x1 - x2) * (x1 - x2) + (y1 - y2) * (y1 - y2));
};

export const circumcircle = (
  { x: x1, y: y1 },
  { x: x2, y: y2 },
  { x: x3, y: y3 }
) => {
  const x12 = x2 - x1;
  const y12 = y2 - y1;
  const x13 = x3 - x1;
  const y13 = y3 - y1;

  const z2 = x12 * (x1 + x2) + y12 * (y1 + y2);
  const z3 = x13 * (x1 + x3) + y13 * (y1 + y3);
  const d = 2.0 * (x12 * (y3 - y2) - y12 * (x3 - x2));

  const x = (y13 * z2 - y12 * z3) / d;
  const y = (x12 * z3 - x13 * z2) / d;

  const r = dist({ x: x1, y: y1 }, { x, y });
  return { x, y, r };
};

export const area = ({ x: x1, y: y1 }, { x: x2, y: y2 }, { x: x3, y: y3 }) => {
  return (x3 * y2 - x2 * y3 - (x3 * y1 - x1 * y3) + (x2 * y1 - x1 * y2)) / 2;
};

export const angle = ({ x: x1, y: y1 }, { x: x2, y: y2 }, { x: x3, y: y3 }) => {
  const a = dist({ x: x2, y: y2 }, { x: x3, y: y3 });
  const b = dist({ x: x1, y: y1 }, { x: x3, y: y3 });
  const c = dist({ x: x1, y: y1 }, { x: x2, y: y2 });

  return (Math.acos((a * a + c * c - b * b) / (2 * a * c)) * 180) / Math.PI;
};

export const incircle = (
  { x: x1, y: y1 },
  { x: x2, y: y2 },
  { x: x3, y: y3 }
) => {
  const a = dist({ x: x2, y: y2 }, { x: x3, y: y3 });
  const b = dist({ x: x1, y: y1 }, { x: x3, y: y3 });
  const c = dist({ x: x1, y: y1 }, { x: x2, y: y2 });
  const p = a + b + c;

  const x = (a * x1 + b * x2 + c * x3) / p;
  const y = (a * y1 + b * y2 + c * y3) / p;

  const r = (1 / 2) * sqrt(((b + c - a) * (c + a - b) * (a + b - c)) / p);
  return { x, y, r };
};

export const orthocenter = (
  { x: x1, y: y1 },
  { x: x2, y: y2 },
  { x: x3, y: y3 }
) => {
  const k1 = (y3 - y2) / (x2 - x3);
  const k2 = (y3 - y1) / (x1 - x3);
  const y = (k1 * y1 - k2 * y2 + x2 - x1) / (k1 - k2);
  const x = k1 * y - k1 * y1 + x1;
  return { x, y };
};

export const centroid = (
  { x: x1, y: y1 },
  { x: x2, y: y2 },
  { x: x3, y: y3 }
) => {
  return { x: (x1 + x2 + x3) / 3, y: (y1 + y2 + y3) / 3 };
};
