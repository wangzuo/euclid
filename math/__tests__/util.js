import { dist, circumcircle, area, angle } from '../util';

test('dist', () => {
  expect(dist({ x: 10, y: 10 }, { x: 120, y: 30 })).toMatchSnapshot();
});

test('circumcircle', () => {
  expect(
    circumcircle(
      { x: 100, y: 100 },
      {
        x: 50,
        y: 200
      },
      {
        x: 300,
        y: 300
      }
    )
  ).toMatchSnapshot();
});

test('area', () => {
  expect(
    area(
      { x: 100, y: 100 },
      {
        x: 50,
        y: 200
      },
      {
        x: 300,
        y: 300
      }
    )
  ).toMatchSnapshot();
});

test('angle', () => {
  expect(
    angle({ x: 100, y: 100 }, { x: 50, y: 200 }, { x: 300, y: 300 })
  ).toMatchSnapshot();
});
