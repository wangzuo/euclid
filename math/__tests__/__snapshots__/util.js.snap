// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`angle 1`] = `85.23635830927383`;

exports[`area 1`] = `15000`;

exports[`circumcircle 1`] = `
Object {
  "r": 141.91155304938667,
  "x": 191.66666666666666,
  "y": 208.33333333333334,
}
`;

exports[`dist 1`] = `111.80339887498948`;
